single_kernel_mongo/managers/mongos_operator.py

```diff
@@ -0,0 +1,318 @@
+        self.share_connection_info()
+
+    @override
+    def on_stop(self) -> None:
+        pass
+
+    @override
+    def start_charm_services(self) -> None:
+        self.workload.start()
+
+    @override
+    def stop_charm_services(self) -> None:
+        self.workload.stop()
+
+    @override
+    def restart_charm_services(self) -> None:
+        self.workload.stop()
+        if not self.state.cluster.config_server_uri:
+            logger.error("Cannot start mongos without a config server db")
+            raise MissingConfigServerError()
+
+        self.mongos_config_manager.set_environment()
+        self.workload.start()
+
+    @override
+    def is_relation_feasible(self, name: str) -> bool:
+        if name != RelationNames.MONGOS_PROXY:
+            return False
+        return True
+
+    def share_connection_info(self):
+        """Shares the connection information of clients."""
+        if not self.state.db_initialised:
+            return
+        if not self.charm.unit.is_leader():
+            return
+        try:
+            self._share_configuration()
+        except PyMongoError as e:
+            raise DeferrableError(f"updating app relation data because of {e}")
+        except ApiError as e:  # Raised for k8s
+            if e.status.code == 404:
+                raise DeferrableError(
+                    "updating app relation data since service not found for more or one units"
+                )
+            raise
+
+    def _share_configuration(self):
+        """Actually shares the configuration according to the substrate."""
+        match self.substrate:
+            case Substrates.VM:
+                if not self.state.mongos_config.password or not self.state.mongos_config.username:
+                    return
+                for relation in self.state.client_relations:
+                    self.mongo_manager.update_app_relation_data_for_config(
+                        relation, self.state.mongos_config
+                    )
+            case Substrates.K8S:
+                for relation in self.state.client_relations:
+                    self.mongo_manager.update_app_relation_data(relation)
+
+    def share_credentials(self, relation: Relation):
+        """Shares credentials to the client."""
+        data_interface = DatabaseProviderData(self.model, relation.name)
+        if not self.charm.unit.is_leader():
+            return
+        new_database_name = data_interface.fetch_relation_field(relation.id, "database")
+        new_extra_user_roles: set[str] = set(
+            json.loads(
+                data_interface.fetch_relation_field(
+                    relation.id,
+                    "extra-user-roles",
+                )
+                or "[]"
+            )
+        )
+        external_connectivity = json.loads(
+            data_interface.fetch_relation_field(relation.id, "external-node-conectivity") or "false"
+        )
```

Comment:
can we retrieve the value for VM for `"external-node-conectivity"` in the state rather than the app databag?